### PR TITLE
Refresh tokens when needed, not on every download

### PR DESCRIPTION
**Pull request by bmbouter**
closes #7643

### CODE REVIEW

File: pulp_ansible/app/downloaders.py [hunk 1]
```diff
@@ -104,12 +141,19 @@ async def _run(self, extra_data=None):
...
         """
-        Update the Bearer token to be used with all requests.
+        Use an existing, or refresh, the Bearer token to be used with all requests.
         """
+        global AUTH_TOKEN
+        global TOKEN_LOCK
+
+        if AUTH_TOKEN:
+            return AUTH_TOKEN
         async with self.TOKEN_LOCK:
```
Comment (fao89) [inline pulp_ansible/app/downloaders.py:153]:
```Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]: pulp: rq.worker:ERROR: Traceback (most recent call last):                                                              
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/local/lib/pulp/lib64/python3.7/site-packages/rq/worker.py", line 936, in perform_job                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     rv = job.perform()                                                                                                 
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/local/lib/pulp/lib64/python3.7/site-packages/rq/job.py", line 684, in perform                             
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     self._result = self._execute()                                                                                     
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/local/lib/pulp/lib64/python3.7/site-packages/rq/job.py", line 690, in _execute                            
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     return self.func(*self.args, **self.kwargs)                                                                        
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 82, in sync                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     d_version.create()                                                                                                 
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulpcore/pulpcore/plugin/stages/declarative_version.py", line 148, in create               
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     loop.run_until_complete(pipeline)                                                                                  
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/lib64/python3.7/asyncio/base_events.py", line 587, in run_until_complete                                  
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     return future.result()                                                                                             
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulpcore/pulpcore/plugin/stages/api.py", line 225, in create_pipeline                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     await asyncio.gather(*futures)                                                                                     
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulpcore/pulpcore/plugin/stages/api.py", line 43, in __call__                              
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     await self.run()                                                                                                   
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 290, in run                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     async for metadata in self._fetch_collections():                                                                   
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 441, in _fetch_collections       
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     count = await _loop_through_pages(not_done)                                                                        
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 399, in _loop_through_pages      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     url = await _get_url(1, versions_url)                                                                              
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 382, in _get_url                 
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     api_endpoint = (await _get_collection_api(root))[0]                                                                
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/tasks/collections.py", line 354, in _get_collection_api      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     api_data = parse_metadata(await downloader.run())                                                                  
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulpcore/pulpcore/download/base.py", line 227, in run                                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     return await self._run(extra_data=extra_data)                                                                      
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/local/lib/pulp/lib64/python3.7/site-packages/backoff/_async.py", line 133, in retry                       
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     ret = await target(*args, **kwargs)                                                                                
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/downloaders.py", line 116, in _run                           
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     return await self._run_with_token_refresh_and_401_retry()                                                          
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/usr/local/lib/pulp/lib64/python3.7/site-packages/backoff/_async.py", line 133, in retry
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     ret = await target(*args, **kwargs)
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/downloaders.py", line 120, in _run_with_token_refresh_and_401_retry                                         
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     token = await self.get_or_update_token()
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:   File "/home/vagrant/devel/pulp_ansible/pulp_ansible/app/downloaders.py", line 153, in get_or_update_token
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]:     async with self.TOKEN_LOCK:
Nov 03 12:58:35 pulp3-source-fedora31.localhost.example.com rq[32090]: AttributeError: 'TokenAuthHttpDownloader' object has no attribute 'TOKEN_LOCK'
```

Comment (bmbouter) [reply pulp_ansible/app/downloaders.py]: fixed on last push, ty

File: pulp_ansible/app/downloaders.py [hunk 2]
```diff
@@ -104,12 +142,19 @@ async def _run(self, extra_data=None):
...
-        async with self.TOKEN_LOCK:
-            log.info("Updating bearer token")
+        global AUTH_TOKEN
+        global TOKEN_LOCK
+
+        if AUTH_TOKEN:
+            return AUTH_TOKEN
+        async with TOKEN_LOCK:
+            if AUTH_TOKEN:
+                return AUTH_TOKEN
```
Comment (fao89) [inline pulp_ansible/app/downloaders.py:156]: I'm trying to understand how the toke is refreshed with these if's

Comment (fao89) [inline pulp_ansible/app/downloaders.py:156]: how the refresh happens even with these ifs?

Comment (fao89) [reply pulp_ansible/app/downloaders.py:156]: actually, I was not giving enough time for the tokens to expire:
```
Nov 03 21:51:00 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:00 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:00 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:00 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:51:01 pulp3-source-fedora31.localhost.example.com gunicorn[36003]: 127.0.0.1 - admin [03/Nov/2020:21:51:01 +0000] "GET /pulp/api/v3/tasks/279b9126-02c6-4f88-a35a-829123583bca/ HTTP/1.1" 200 1237 "-" "OpenAPI-Generator//python"
Nov 03 21:51:02 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
.
.
.
Nov 03 21:53:50 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:53:50 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:53:50 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:53:50 pulp3-source-fedora31.localhost.example.com rq[35999]: pulp: backoff:INFO: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
Nov 03 21:53:50 pulp3-source-fedora31.localhost.example.com rq[35999]: Backing off _run_with_token_refresh_and_401_retry(...) for 0.0s (aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://cloud.redhat.com/api/automation-hub/v3/collections/check_point/mgmt/versions/?offset=0&limit=100'))
```

diff:
```diff
diff --git a/pulp_ansible/app/tasks/collections.py b/pulp_ansible/app/tasks/collections.py
index 3e1db78..342ff81 100644
--- a/pulp_ansible/app/tasks/collections.py
+++ b/pulp_ansible/app/tasks/collections.py
@@ -4,6 +4,7 @@ from gettext import gettext as _
 import json
 import logging
 import math
+import time
 import tarfile
 from urllib.parse import urljoin, urlparse, urlunparse
 
@@ -408,6 +409,7 @@ class CollectionSyncFirstStage(Stage):
                 page_count = math.ceil(float(count) / float(PAGE_SIZE))
 
             for page in range(1, page_count + 1):
+                time.sleep(120)
                 url = await _get_url(page, versions_url)
                 downloader = remote.get_downloader(url=url)
                 not_done.add(downloader.run())
diff --git a/pulp_ansible/tests/functional/cli/test_collection_install.py b/pulp_ansible/tests/functional/cli/test_collection_install.py
index c40af10..c8bb3f6 100644
--- a/pulp_ansible/tests/functional/cli/test_collection_install.py
+++ b/pulp_ansible/tests/functional/cli/test_collection_install.py
@@ -94,7 +94,6 @@ class InstallCollectionTestCase(unittest.TestCase):
         """Test whether ansible-galaxy can install a Collection hosted by Pulp."""
         body = gen_ansible_remote(
             url=AUTOMATION_HUB_URL,
-            requirements_file="collections:\n  - ansible.posix",
             auth_url=AH_AUTH_URL,
             token=os.environ["AUTOMATION_HUB_TOKEN_AUTH"],
             tls_validation=False,
```

File: pulp_ansible/app/downloaders.py [hunk 3]
```diff
@@ -90,11 +94,46 @@ async def _run(self, extra_data=None):
...
+                 more information.
+        """
+        while True:
+            token = await self.get_or_update_token()
             # Keycloak Token
             headers = {"Authorization": "Bearer {token}".format(token=token)}
+            try:
+                return await self._run_with_additional_headers(headers)
+            except ClientResponseError as exc:
+                if exc.status is 401:
```
Comment (fao89) [inline pulp_ansible/app/downloaders.py:119]:
```suggestion
                if exc.status == 401:
```

Comment (bmbouter) [reply pulp_ansible/app/downloaders.py:119]: fixing in next push along w/ CI. Thanks!

**Review by fao89 — APPROVED** [review]
Thank you Brian!
Sorry for all this testing back and forth

**Pull request merged**